lib/fizzy/instantiate.hpp

```diff
@@ -21,23 +21,81 @@ namespace fizzy
 struct ExecutionResult;
 struct Instance;
 
-/// Function representing WebAssembly or host function execution.
-using execute_function = std::function<ExecutionResult(Instance&, const Value*, int depth)>;
+/// Function pointer to the execution function.
+using ExecuteFunctionPtr = ExecutionResult (*)(
+    void* context, Instance&, const Value*, int depth) noexcept;
+
+/// Function representing either WebAssembly or host function execution.
+class ExecuteFunction
+{
+    /// Pointer to WebAssembly function instance.
+    /// Equals nullptr in case this ExecuteFunction represents host function.
+    Instance* m_instance = nullptr;
+
+    /// Index of WebAssembly function.
+    /// Equal 0 in case this ExecuteFunction represents host function.
+    FuncIdx m_func_idx = 0;
+
+    /// Pointer to a host function.
+    /// Equals nullptr in case this ExecuteFunction represents WebAssembly function.
+    ExecuteFunctionPtr m_host_function = nullptr;
+
+    /// Opaque context of host function execution, which is passed to it as a context parameter.
+    /// `shared_ptr` is used because ExecuteFunction can be copied.
+    /// Empty in case this ExecuteFunction represents WebAssembly function.
+    std::shared_ptr<void> m_host_context;
+
+public:
+    /// Default function constructor.
+    ExecuteFunction() noexcept = default;
+    /// Copy constructor.
+    ExecuteFunction(const ExecuteFunction&) noexcept = default;
+    /// Move constructor.
+    ExecuteFunction(ExecuteFunction&&) noexcept = default;
+    /// Copy assignment operator.
+    ExecuteFunction& operator=(const ExecuteFunction&) noexcept = default;
+    /// Move assignment operator.
+    ExecuteFunction& operator=(ExecuteFunction&&) noexcept = default;
+
+    // WebAssembly function constructor.
+    ExecuteFunction(Instance& instance, FuncIdx func_idx) noexcept
+      : m_instance{&instance}, m_func_idx{func_idx}
+    {}
+
+    /// Host function constructor without taking ownership of context.
+    /// Context lifetime is assumed to be managed by the caller.
+    ExecuteFunction(ExecuteFunctionPtr f, void* context = nullptr)
+      : m_host_function{f}, m_host_context{context, [](void*) noexcept {}}
```

Comment:
It might be not a great idea - I think this shared pointer will still have an overhead of reference counting. Maybe better solution would be to have a union of raw pointer and `shared_pointer` and use raw pointer here.

But also this contstructor is currently used only with `nullptr` contexr, I just added it at some point as a general version because it seemed it would be convenient for a user (still seems so). I think I'll leave this feature for a later PR.